src/oemof/solph/_energy_system.py

```diff
@@ -30,14 +35,150 @@ class EnergySystem(es.EnergySystem):
+                raise AttributeError(msg)
+
+            timeindex = timeindex.union(
+                pd.date_range(
+                    timeindex[-1] + timeindex.freq,
+                    periods=1,
+                    freq=timeindex.freq,
+                )
+            )
+
+        # catch wrong combinations and infer timeincrement from timeindex.
+        if timeincrement is not None and timeindex is not None:
+            msg = (
+                "Specifying the timeincrement and the timeindex parameter at "
+                "the same time is not allowed since these might be "
+                "conflicting to each other."
+            )
+            raise AttributeError(msg)
+
+        elif timeindex is not None and timeincrement is None:
+            df = pd.DataFrame(timeindex)
+            timedelta = df.diff()
+            timeincrement = pd.Series(
+                (timedelta / np.timedelta64(1, "h"))[1:].set_index(0).index
+            )
+
+        if timeincrement is not None and (pd.Series(timeincrement) <= 0).any():
+            msg = (
+                "The time increment is inconsistent. Negative values and zero "
+                "is not allowed.\nThis is caused by a inconsistent "
+                "timeincrement parameter or an incorrect timeindex."
+            )
+            raise TypeError(msg)
+
+        super().__init__(
+            timeindex=timeindex, timeincrement=timeincrement, **kwargs
+        )
+
+
+def create_year_index(year, interval=1, number=None):
+    """
+    Create a datetime index for one year.
+
+    Notes
+    -----
+    To create 8760 hourly intervals for a non leap year a datetime index with
+    8761 time points need to be created. So the number of time steps is always
+    the number of intervals plus one.
+
+    Parameters
+    ----------
+    year : int
+        The year of the index.
+    interval : float
+        The time interval in hours e.g. 0.5 for 30min or 2 for a two hour
+        interval (default: 1).
+    number : int
+        The number of time intervals. By default number is calculated to create
+        an index of one year. For a shorter or longer period the number of
+        intervals can be set by the user.
+
+    Examples
+    --------
+    >>> len(create_year_index(2014))
+    8761
+    >>> len(create_year_index(2012))  # leap year
+    8785
+    >>> len(create_year_index(2014, interval=0.5))
+    17521
+    >>> len(create_year_index(2014, interval=0.5, number=10))
+    11
+    >>> len(create_year_index(2014, number=10))
+    11
+    >>> str(create_year_index(2014, interval=0.5, number=10)[-1])
+    '2014-01-01 05:00:00'
+    >>> str(create_year_index(2014, interval=2, number=10)[-1])
+    '2014-01-01 20:00:00'
+    """
+    if number is None:
+        if calendar.isleap(year):
+            hoy = 8784
+        else:
+            hoy = 8760
+        number = hoy / interval
+    return pd.date_range(
+        f"1/1/{year}", periods=number + 1, freq=f"{interval}H"
+    )
```

Comment:
```suggestion
        number = round(hoy / interval)
    return pd.date_range(
        f"1/1/{year}", periods=number + 1, freq=f"{interval}H"
    )
```

`pd.date_range` expects an integer for the number of periods. We should round so that e.g. 0.99 is not truncated.